mender-update/cli/cli_test.cpp

```diff
@@ -1762,3 +1766,477 @@ rootfs-image.checksum=f2ca1bb6c7e907d06dafe4687e579fce76b37e4e93b7605022da52e6cc
+				{"ArtifactCommit_Leave_01", ExitCode {0}},
+			},
+		.expected = R"(Download_Enter_01
+Download
+Download_Leave_01
+ArtifactInstall_Enter_01
+ArtifactInstall
+ArtifactInstall_Leave_01
+NeedsArtifactReboot
+SupportsRollback
+ArtifactCommit_Enter_01
+ArtifactCommit
+Cleanup
+ArtifactCommit_Leave_01
+)",
+	},
+	{
+		.case_name = "commit_enter_error",
+		.expected_exit_code = 1,
+		.scripts =
+			{
+				{"Download_Enter_01", ExitCode {0}},
+				{"Download_Leave_01", ExitCode {0}},
+				{"ArtifactInstall_Enter_01", ExitCode {0}},
+				{"ArtifactInstall_Enter_02", ExitCode {0}},
+				{"ArtifactInstall_Leave_01", ExitCode {0}},
+				{"ArtifactCommit_Enter_01", ExitCode {1}},
+				{"ArtifactCommit_Error_01", ExitCode {0}},
+				{"ArtifactCommit_Error_02", ExitCode {0}},
+				{"ArtifactCommit_Leave_01", ExitCode {0}},
+				{"ArtifactFailure_Enter_01", ExitCode {0}},
+				{"ArtifactFailure_Leave_01", ExitCode {0}},
+			},
+
+		.expected = R"(Download_Enter_01
+Download
+Download_Leave_01
+ArtifactInstall_Enter_01
+ArtifactInstall_Enter_02
+ArtifactInstall
+ArtifactInstall_Leave_01
+NeedsArtifactReboot
+SupportsRollback
+ArtifactCommit_Enter_01
+ArtifactCommit_Error_01
+ArtifactCommit_Error_02
+SupportsRollback
+ArtifactFailure_Enter_01
+ArtifactFailure
+Cleanup
+ArtifactFailure_Leave_01
+)",
+	},
+	{
+		.case_name = "commit_leave_error",
+		.expected_exit_code = 0,
+		.scripts =
+			{
+				{"Download_Enter_01", ExitCode {0}},
+				{"Download_Leave_01", ExitCode {0}},
+				{"Download_Error_01", ExitCode {0}},
+				{"Download_Error_02", ExitCode {0}},
+				{"ArtifactInstall_Enter_01", ExitCode {0}},
+				{"ArtifactInstall_Leave_01", ExitCode {0}},
+				{"ArtifactInstall_Leave_02", ExitCode {0}},
+				{"ArtifactCommit_Enter_01", ExitCode {0}},
+				{"ArtifactCommit_Leave_01", ExitCode {1}},
+				{"ArtifactCommit_Error_01", ExitCode {0}}, // Should never run
+				{"ArtifactFailure_Enter_01", ExitCode {0}},
+				{"ArtifactFailure_Leave_01", ExitCode {0}},
+			},
+		.expected = R"(Download_Enter_01
+Download
+Download_Leave_01
+ArtifactInstall_Enter_01
+ArtifactInstall
+ArtifactInstall_Leave_01
+ArtifactInstall_Leave_02
+NeedsArtifactReboot
+SupportsRollback
+ArtifactCommit_Enter_01
+ArtifactCommit
+Cleanup
+ArtifactCommit_Leave_01
+)",
+	}};
+
+vector<StandaloneStateScriptTestCase> standalone_failure_script_test_cases {
+	{
+		.case_name = "failure_enter_error",
+		.expected_exit_code = 1,
+		.scripts =
+			{
+				{"Download_Enter_01", ExitCode {0}},
+				{"Download_Leave_01", ExitCode {0}},
+				{"ArtifactInstall_Enter_01", ExitCode {0}},
+				{"ArtifactInstall_Enter_02", ExitCode {0}},
+				{"ArtifactInstall_Leave_01", ExitCode {0}},
+				{"ArtifactCommit_Enter_01", ExitCode {1}}, // Trigger failure
+				{"ArtifactCommit_Leave_01", ExitCode {0}},
+				{"ArtifactFailure_Enter_01", ExitCode {1}}, // Should not matter
+				{"ArtifactFailure_Enter_02", ExitCode {1}}, // Should not matter
+				{"ArtifactFailure_Leave_01", ExitCode {1}}, // Should not matter
```

Comment:
Indeed. But this is not reflected in your test data below. Only the first of each `Enter` / `Leave` category is run.

I actually had to run this with the old client to be sure, since the documentation is a bit ambiguous on this point, but I confirmed it: For `Enter` and `Leave` scripts in the failure categories, every script is always run, even if the first script returns error. So you probably need to decouple the "run every script" logic from the `Error` category, and make it a configurable flag on the runner instead.

Reply:
True - good point!